src/View/ViewLayerData.php

```diff
@@ -0,0 +1,254 @@
+<?php
+
+namespace SilverStripe\View;
+
+use BadMethodCallException;
+use InvalidArgumentException;
+use IteratorAggregate;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Model\ModelData;
+use SilverStripe\Model\ModelDataCustomised;
+use SilverStripe\ORM\FieldType\DBClassName;
+use Stringable;
+use Traversable;
+
+class ViewLayerData implements IteratorAggregate, Stringable
+{
+    use Injectable;
+
+    public const TYPE_PROPERTY = 'property';
+
+    public const TYPE_METHOD = 'method';
+
+    public const TYPE_ANY = 'any';
+
+    /**
+     * Special variable names that can be used to get metadata about values
+     */
+    public const META_DATA_NAMES = [
+        // Gets a DBClassName with the class name of $this->data
+        'ClassName',
+        // Returns $this->data
+        'Me',
+    ];
+
+    private object $data;
+
+    public function __construct(mixed $data, mixed $source = null, string $name = '')
+    {
+        if ($data === null) {
+            throw new InvalidArgumentException('$data must not be null');
+        }
+        if ($data instanceof ViewLayerData) {
+            $data = $data->data;
+        } else {
+            $source = $source instanceof ModelData ? $source : null;
+            $data = CastingService::singleton()->cast($data, $source, $name);
+        }
+        $this->data = $data;
+    }
+
+    /**
+     * Needed so we can rewind in SSViewer_Scope::next() after getting itemIteratorTotal without throwing an exception.
+     */
+    public function getIteratorCount(): int
+    {
+        $count = $this->getRawDataValue('count');
+        if (is_numeric($count)) {
+            return $count;
+        }
```

Comment:
Try to get a property/method/getter first - e.g. `DataList::Count()` has some special logic in it.